pennylane/hf/matrices.py

```diff
@@ -0,0 +1,343 @@
+        return v
+
+    return attraction
+
+
+def repulsion_tensor(basis_functions):
+    r"""Return a function that computes the electron repulsion tensor for a given set of basis
+    functions.
+
+    Args:
+        basis_functions (list[BasisFunction]): basis functions
+
+    Returns:
+        function: function that computes the electron repulsion tensor
+
+    **Example**
+
+    >>> symbols  = ['H', 'H']
+    >>> geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]], requires_grad = False)
+    >>> alpha = np.array([[3.42525091, 0.62391373, 0.1688554],
+    >>>                   [3.42525091, 0.62391373, 0.1688554]], requires_grad=True),
+    >>> mol = Molecule(symbols, geometry, alpha=alpha)
+    >>> args = [alpha]
+    >>> repulsion_tensor(mol.basis_set)(*args)
+    array([[[[0.77460595, 0.56886144], [0.56886144, 0.65017747]],
+            [[0.56886144, 0.45590152], [0.45590152, 0.56886144]]],
+           [[[0.56886144, 0.45590152], [0.45590152, 0.56886144]],
+            [[0.65017747, 0.56886144],[0.56886144, 0.77460595]]]])
+    """
+
+    def repulsion(*args):
+        r"""Construct the electron repulsion tensor for a given set of basis functions.
+
+        Args:
+            args (array[float]): initial values of the differentiable parameters
+
+        Returns:
+            array[float]: the electron repulsion tensor
+        """
+        n = len(basis_functions)
+        e = anp.zeros((n, n, n, n))
+        e_calc = []
+
+        for i, a in enumerate(basis_functions):
+            for j, b in enumerate(basis_functions):
+                for k, c in enumerate(basis_functions):
+                    for l, d in enumerate(basis_functions):
+
+                        if [i, j, k, l] not in e_calc:
+                            if args:
+                                args_abcd = []
+                                for m in range(len(args)):
+                                    args_abcd.append(args[m][[i, j, k, l]])
+                                repulsion_integral = generate_repulsion(a, b, c, d)(*args_abcd)
+                            else:
+                                repulsion_integral = generate_repulsion(a, b, c, d)()
+
+                            o = anp.zeros((n, n, n, n))
+                            o[i, j, k, l] = o[k, l, i, j] = o[j, i, l, k] = o[l, k, j, i] = 1.0
+                            o[j, i, k, l] = o[l, k, i, j] = o[i, j, l, k] = o[k, l, j, i] = 1.0
+                            e = e + repulsion_integral * o
+                            e_calc = e_calc + [
+                                [i, j, k, l],
+                                [k, l, i, j],
+                                [j, i, l, k],
+                                [l, k, j, i],
+                                [j, i, k, l],
+                                [l, k, i, j],
+                                [i, j, l, k],
+                                [k, l, j, i],
+                            ]
```

Comment:
is this list keeping track of the two-electron integral symmetries?

Reply:
Yes.